src/lib/core/datetime.c

```diff
@@ -306,13 +312,23 @@ denormalize_interval_nsec(int64_t *psec, int *pnsec)
 	if (nsec == 0)
 		return;
 
-	bool zero_or_same_sign = sec == 0 || (sec < 0) == (nsec < 0);
-	if (zero_or_same_sign && labs(nsec) < NANOS_PER_SEC)
+	if (zero_or_same_sign(sec, nsec) && labs(nsec) < NANOS_PER_SEC)
 		return;
+	sec += nsec / NANOS_PER_SEC;
+	nsec %= NANOS_PER_SEC;
 
-	int64_t normalized_nsec = sec * NANOS_PER_SEC + nsec;
-	sec = normalized_nsec / NANOS_PER_SEC;
-	nsec = normalized_nsec % NANOS_PER_SEC;
+	/* if we happen to receive different signs of sec and nsec
+	 * after all these manipulations, then normalize their sign
+	 */
+	if (!zero_or_same_sign(sec, nsec)) {
+		if (nsec < 0) {
+			sec--;
+			nsec = NANOS_PER_SEC + nsec;
+		} else {
+			sec++;
+			nsec = -NANOS_PER_SEC + nsec;
+		}
+	}
```

Comment:
I doubt my solution is any better than yours, but I still think it looks a tiny bit simpler.
I couldn't think of anything better, so feel free to leave your version, if you like it more.
```diff
diff --git a/src/lib/core/datetime.c b/src/lib/core/datetime.c
index 5c19c8cbe..b17d2cf68 100644
--- a/src/lib/core/datetime.c
+++ b/src/lib/core/datetime.c
@@ -314,20 +314,16 @@ denormalize_interval_nsec(int64_t *psec, int *pnsec)
 
        if (zero_or_same_sign(sec, nsec) && labs(nsec) < NANOS_PER_SEC)
                return;
-       sec += nsec / NANOS_PER_SEC;
-       nsec %= NANOS_PER_SEC;
+       sec += DIV(nsec, NANOS_PER_SEC);
+       nsec = MOD(nsec, NANOS_PER_SEC);
 
-       /* if we happen to receive different signs of sec and nsec
-        * after all these manipulations, then normalize their sign
+       /*
+        * We crafted a positive nsec value, so convert it to negative, if secs
+        * are also negative.
         */
-       if (!zero_or_same_sign(sec, nsec)) {
-               if (nsec < 0) {
-                       sec--;
-                       nsec = NANOS_PER_SEC + nsec;
-               } else {
-                       sec++;
-                       nsec = -NANOS_PER_SEC + nsec;
-               }
+       if (sec < 0) {
+               sec++;
+               nsec -= NANOS_PER_SEC;
        }
 
        *psec = sec;
```